src/webnn/native/dml/BackendDML.cpp

```diff
@@ -0,0 +1,143 @@
+                                                               IID_PPV_ARGS(&adapter)) !=
+                       DXGI_ERROR_NOT_FOUND) {
+                    DXGI_ADAPTER_DESC1 pDesc;
+                    adapter->GetDesc1(&pDesc);
+                    // An adapter called the "Microsoft Basic Render Driver" is always present.
+                    // This adapter is a render-only device that has no display outputs. See here
+                    // for documentation on filtering WARP adapter:
+                    // https://docs.microsoft.com/en-us/windows/desktop/direct3ddxgi/d3d10-graphics-programming-guide-dxgi#new-info-about-enumerating-adapters-for-windows-8
+                    bool isSoftwareAdapter = pDesc.Flags == DXGI_ADAPTER_FLAG_SOFTWARE ||
+                                             (pDesc.VendorId == 0x1414 && pDesc.DeviceId == 0x8c);
+                    if (!isSoftwareAdapter) {
+                        break;
+                    }
+                }
+            } else {
+                WEBNN_RETURN_IF_FAILED(dxgiFactory->EnumWarpAdapter(IID_PPV_ARGS(&adapter)));
+            }
+            return S_OK;
+        }
+
+    }  // namespace
+
+    Backend::Backend(InstanceBase* instance)
+        : BackendConnection(instance, wnn::BackendType::DirectML) {
+    }
+
+    MaybeError Backend::Initialize() {
+        return {};
+    }
+
+    ContextBase* Backend::CreateContext(ContextOptions const* options) {
+        wnn::DevicePreference devicePreference =
+            options == nullptr ? wnn::DevicePreference::Default : options->devicePreference;
+        bool useGpu = devicePreference == wnn::DevicePreference::Cpu ? false : true;
+        DXGI_GPU_PREFERENCE gpuPreference = DXGI_GPU_PREFERENCE_UNSPECIFIED;
+        wnn::PowerPreference powerPreference =
+            options == nullptr ? wnn::PowerPreference::Default : options->powerPreference;
+        switch (powerPreference) {
+            case wnn::PowerPreference::High_performance:
+                gpuPreference = DXGI_GPU_PREFERENCE::DXGI_GPU_PREFERENCE_HIGH_PERFORMANCE;
+                break;
+            case wnn::PowerPreference::Low_power:
+                gpuPreference = DXGI_GPU_PREFERENCE::DXGI_GPU_PREFERENCE_MINIMUM_POWER;
+                break;
+            default:
+                break;
+        }
+
+        bool useDebugLayer = false;
+#ifdef _DEBUG
+        useDebugLayer = true;
+#endif
+        if (useDebugLayer) {
+            ComPtr<ID3D12Debug> debug;
+            if (SUCCEEDED(D3D12GetDebugInterface(IID_PPV_ARGS(&debug)))) {
+                debug->EnableDebugLayer();
+            }
+        }
+
+        ComPtr<IDXGIAdapter1> adapter;
+        if (FAILED(EnumAdapter(gpuPreference, useGpu, adapter))) {
+            dawn::ErrorLog() << "Failed to enumerate adapters for creating the context.";
+            return nullptr;
+        }
+
+        // Create the D3D device.
+        if (FAILED(D3D12CreateDevice(adapter.Get(), D3D_FEATURE_LEVEL_11_0,

```

Comment:
As we disscusion, the device and command queue can't be created when creating context, they should be created when connecting DML backend.